File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    let res = [];
    nums.sort((a, b) => a - b);
    let map = {};

    //hashmap to store last index of each item
    for (let i = 0; i < nums.length; i++) {
        map[nums[i]] = i;
    }

    //using 2 pointer and hashmap

    for (let i = 0; i < nums.length; i++) {
        //skip duplicates
        if (i > 0 && nums[i] == nums[i - 1]) {
            continue;
        }

        //for valid i
        let j = i + 1;

        while (j < nums.length) {
            //skip duplicates
            //here j != i+1 is important else it will be skipping cases like [-1, -1, 2] because i and i+1 are same ones
            if (j != i + 1 && nums[j] == nums[j - 1]) {
                j++;
                continue;
            }

            let key = -(nums[i] + nums[j]);
            if (key in map) {
                //to skip the duplicates
                if (map[key] > j) {
                    res.push([nums[i], nums[j], key]);
                }
            }

            j++;
        }
    }

    return res;
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        n = len(nums)
        res = []
        for i in range(n):
            # skip duplicates
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j = i + 1
            k = n - 1

            while j < k:
                if nums[i] + nums[j] + nums[k] == 0:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    # skip duplicates
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                elif nums[i] + nums[j] + nums[k] < 0:
                    j += 1
                else:
                    k -= 1

        return res
